pkg/dm/message_impl.go

```diff
@@ -0,0 +1,83 @@
+package dm
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/errors"
+
+	"github.com/hanfei1991/microcosm/pkg/p2p"
+)
+
+type MessageIDAllocator struct {
+	mu sync.Mutex
+	id uint64
+}
+
+func (a *MessageIDAllocator) Alloc() uint64 {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.id++
+	return a.id
+}
+
+type Sender interface {
+	SendMessage(ctx context.Context, topic p2p.Topic, message interface{}, nonblocking bool) error
+}
+
+// MessageImpl implement a simple synchronous request/response message pattern since the lib currently only support asynchronous message.
+// Caller should persist the request message if needed.
+// Caller should add retry mechanism if needed.
+type MessageImpl struct {
```

Comment:
maybe MessageRecorder, MessageRouter or something. Message makes me think of a message entity that has NewMessage and some (de)serialization method.

Reply:
I change it to MessagePair...